src/odemis/driver/technolution.py

```diff
@@ -659,100 +672,151 @@ def _getScanStepSizeBits(self):
         A single sawtooth profile (rise and crash) followed by a flyback period (X=0) is used as trajectory. No
         smoothing or low-pas filtering is used to create the trajectory of these setpoints.
 
-        :return: List of ints holding the setpoints in X direction.
+        :return (List of ints): Setpoints in x direction.
         """
-        descan_period = self.clockPeriod.value  # in seconds
-        dwellTime = self.parent._ebeam_scanner.dwellTime.value  # in seconds
-        X_cell_resolution = self.parent._mppc.cellCompleteResolution.value[0]  # pixels
-        scan_offset_bits = convertBits(self.scanOffset.value[0], numpy.array(self.scanOffset.range)[:, 1])
-        scan_gain_bits = convertBits(self.scanGain.value[0], numpy.array(self.scanGain.range)[:, 1])
-
-        # Scanning points are determined via the linear equation x_setpoints = A*t + B
+        # The setpoints for an acquisition resemble a linear ramp for scanning a line of pixels, followed by the
+        # flyback, where the descanner moves back to its starting position.
+        # The scanning setpoints are determined via the linear equation x_setpoints = A*t + B
         # With 'A' being the slope of the scanning line which is: (maximum - minimum) / scanning time
         # And 'B' being the offset
 
+        descan_period = self.clockPeriod.value  # in seconds
+        dwellTime = self.parent._ebeam_scanner.dwellTime.value  # in seconds
+        x_cell_resolution = self.parent._mppc.cellCompleteResolution.value[0]  # pixels
+        scan_offset_bits = convert2Bits(self.scanOffset.value[0], numpy.array(self.scanOffset.range)[:, 1])
+        scan_gain_bits = convert2Bits(self.scanGain.value[0], numpy.array(self.scanGain.range)[:, 1])
+
         # all units in seconds
-        scanning_time = dwellTime * X_cell_resolution
+        scanning_time = dwellTime * x_cell_resolution
 
-        # Scanning time which does not fall in a multiple of whole descan periods
+        # Remainder of the scanning time which is not a whole multiple of the descan periods.
         remainder_scanning_time = scanning_time % descan_period
 
-        # Use timing setpoints because the maximum is defined by the time scale.
-        time_setpoints_scan = numpy.linspace(0, scanning_time - remainder_scanning_time,
-                                             int(scanning_time // descan_period))
-        # TODO K.K. check precision of descan period on ASM with Robert
+        # Calculate the time stamps for the setpoints excluding the remainder of the scanning time.
+        number_setpoints = scanning_time // descan_period
+        timestamp_setpoints = numpy.linspace(0, scanning_time - remainder_scanning_time, number_setpoints)
+
         # An absolute tolerance of 1e-6 is one order of magnitude more precision than the descan period.
+        # Use almost_equal to handle floating point errors.
         if not all(almost_equal(time_step, descan_period, rtol=0, atol=1e-6) for time_step in
```

Comment:
The calculation of the setpoints should always work. I don't currently see how it can fail. This check looks better in the test cases I would say. Not sure it is a good idea to fail here. @pieleric What do you think. Is it ok to fail in the calculation of the setpoints for an acquisition? And if yes, how should the user deal with it?

Reply:
IMHO, it *sometimes* makes sense to add checks at runtime, but under the following two conditions:
* The performance degradation should be minimal (<10% from the actual function)
* If a failure is detected, log the heck of data, and try to recover.

So here, it could make sense, but under the condition that instead of raising an error, you log it (at the error level), with every input of the function, so that it's possible to reproduce the issue. Also, in this case, I'd suggest to still continue the function. For the other cases, like checking the data range, in case of error, clip the data, and continue.

Still, the best is to have the function be well tested in the unit tests. For making the testing easier, you could maybe have the core of the function as a separate `@static_method`. It wouldn't rely on any attribute of the `self` (and friends), but instead it'd take all its parameters as... parameters, and it return a numpy array of the setpoints. The parameters would be the "core" values for a curve generation, like "number_setpoints", "number_flyback_points", "slope", "offset".

Reply:
@Sabrina85 was right we cannot reach this error, numpy.linspace should evenly space, so basically we are here testing numpy. Therefore, the error can be removed.